Proper support for namespaces and other enhancement/fixes

neikeq opened the pull request:
- Fix internal assembly load from (it wasn't actually loading from the specified path) and make sure we load API assemblies from `res://.mono/assemblies/` instead of other locations.
- Support globs in csproj includes.
- Do not generate API project GUIDs randomly.
- Fix msvc warnings in mono module.
- Proper namespace support: Parse C# script namespace and class. Fixes #17409

The parser to extract the class name and namespace is an improved version of the one uploaded by @reduz (4613cb7874a7a6beb3195bebf06d7ecfbd2454c3), with a few modifications and fixes.
If this turns not to be a reliable solution, we can always switch to Microsoft.CodeAnalysis.CSharp.

**DNM:** This PR adds a tiny dependency for the editor (https://github.com/dazinator/DotNet.Glob), and I'm still trying to figure out where the license info should go.

neikeq committed:
Fix internal assembly load from

- Also make sure we load API assemblies from 'res://.mono/assemblies/'.

exts commented:
Building on windows 10 failed for me:

```
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error : Unable to load the service index for sour
ce https://api.nuget.org/v3/index.json. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTool
s.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   An error occurred while sending the req
uest [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Error: TrustFailure (The authentication
 or decryption has failed.) [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   The authentication or decryption has fa
iled. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Invalid certificate received from serve
r. Error code: 0xffffffff800b0109 [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.cspr
oj]
Done Building Project "D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj" (Restore
 target(s)) -- FAILED.

Done Building Project "D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.sln" (Restore;Bu
ild target(s)) -- FAILED.


Build FAILED.

"D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.sln" (Restore;Build target) (1) ->
"D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj" (Restore target) (2) ->
(Restore target) ->
  C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error : Unable to load the service index for so
urce https://api.nuget.org/v3/index.json. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTo
ols.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   An error occurred while sending the req
uest [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Error: TrustFailure (The authentication
 or decryption has failed.) [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   The authentication or decryption has fa
iled. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Invalid certificate received from serve
r. Error code: 0xffffffff800b0109 [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.cspr
oj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:02.26
scons: *** [bin\GodotSharp\Tools\GodotSharpTools.dll] RuntimeError : GodotSharpTools build failed
Traceback (most recent call last):
  File "C:\Python27\Scripts\..\Lib\site-packages\scons-3.0.0\SCons\Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "D:\godot-master\build\godot\modules\mono\SCsub", line 231, in mono_build_solution
    raise RuntimeError('GodotSharpTools build failed')
RuntimeError: GodotSharpTools build failed
   Creating library bin\godot.windows.tools.64.mono.lib and object bin\godot.windows.tools.64.mono.exp
scons: building terminated because of errors.
```

akien-mga commented on modules/mono/editor/GodotSharpTools/Project/ProjectGenerator.cs:
```diff
@@ -6,6 +6,9 @@ namespace GodotSharpTools.Project
     public static class ProjectGenerator
     {
+        const string CoreApiProjectGuid = "{AEBF0036-DA76-4341-B651-A3F2856AB2FA}";
+        const string EditorApiProjectGuid = "{8FBEC238-D944-4074-8548-B3B524305905}";
```
Will those always stay the same? Might be worth documenting where those IDs come from/what they mean with a short comment.

neikeq replied:
These are randomly generated unique IDs. They will stay the same, unlike before when we were generating them every time for the same project.

akien-mga commented:
> This PR adds a tiny dependency for the editor (https://github.com/dazinator/DotNet.Glob), and I'm still trying to figure out where the license info should go.

I don't see the dependency added in the `thirdparty` folder, is it downloaded at runtime? Usually the process should be to put the source files in `thirdparty`, document their provenance and license in `thirdparty/README.md` and add licensing details in `COPYRIGHT.txt`.

But if it's a dependency downloaded as a binary blob at compile time, I'm not sure how to handle that (apart from saying "Eeew" because I hate buildsystems that rely on downloading stuff from the Internet :P). I guess we should design a way for the Mono module to register new thirdparty components to add to the Engine singleton's license details (currently parsed from `COPYRIGHT.txt` at compile time).

neikeq commented:
Yes, it's downloaded at build time. The NuGet reference is here: https://github.com/godotengine/godot/pull/23162/files#diff-15a6ab61c9cc330d83e32426b12e15c6R34

neikeq committed:
Support globs in csproj includes

neikeq committed:
Do not generate API project GUIDs randomly

neikeq committed:
Fix msvc warnings in mono module

- `modules\mono\csharp_script.cpp(576): warning C4099: 'CSharpScriptDepSort': type name first seen using 'class' now seen using 'struct'`
- `modules\mono\signal_awaiter_utils.cpp(144): warning C4003: not enough actual parameters for macro 'ERR_FAIL_V'`
- `modules\mono\editor\net_solution.cpp(101): warning C4129: '%': unrecognized character escape sequence`
- (several) `modules\mono\glue\cs_compressed.gen.h(222): warning C4129: 'E': unrecognized character escape sequence`

neikeq committed:
Parse C# script namespace and class

- Added a very simple parser that can extract the namespace and class name of a C# script.

exts commented:
[NSTest2.zip](https://github.com/godotengine/godot/files/2502677/NSTest2.zip)

Works flawlessly, awesome :)

exts commented:
@neikeq#7308 

```
mono_gc_handle.cpp
modules\mono\csharp_script.cpp(835): error C2065: 'DEBUG_ENABLED': undeclared identifier
modules\mono\csharp_script.cpp(843): error C3861: 'read_all_file_utf8': identifier not found
modules\mono\csharp_script.cpp(2608): error C3861: 'read_all_file_utf8': identifier not found
scons: *** [modules\mono\csharp_script.windows.opt.64.obj] Error 2
scons: building terminated because of errors.
```

`scons p=windows target=release tools=no module_mono_enabled=yes -j3`

When building export templates

neikeq commented:
@exts Fixed

aaronfranke commented:
Previous commit (dbaa22329761ce7a4d8bab291871b5b5ba359cbe) builds fine.

After this commit (d47cec43f2f4ef2cd6d1a0acf6e99a9dd8d31eef):

```
scons: *** [bin/GodotSharp/Tools/GodotSharpTools.dll] RuntimeError : Cannot find NuGet executable
Traceback (most recent call last):
  File "/usr/lib/scons/SCons/Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/home/aaronfranke/workspace/godot/modules/mono/SCsub", line 291, in mono_build_solution
    raise RuntimeError('Cannot find NuGet executable')
```

aaronfranke commented:
When I compile, it generates several DLL files which are not in the repo and are not in `.gitignore`.

KellyThomas commented:
@hedgehog90 If you don't have certs configured yet nuget you will need to do something like this:  
https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed/#49799509

hedgehog90 commented:
@KellyThomas Thanks! That did the trick.

neikeq commented:
Docs updated: https://github.com/godotengine/godot-docs/pull/1911

aaronfranke commented:
Exporting doesn't work for me with 3.1 Alpha 2:

```
$ ERROR: connect: Signal 'confirmed' is already connected to given method '_confirmed' in that object.
   At: core/object.cpp:1463.
ERROR: load_from: FATAL: Condition ' !FileAccess::exists(p_path) ' is true.
   At: modules/mono/mono_gd/gd_mono_assembly.cpp:466.
Stacktrace:
```

exts commented:
@aaronfranke https://github.com/godotengine/godot/commit/9fa1e0b4153102913cddc6380207f5c43e6ebaac make sure you have that fix
